dc/249 stream api

cdawei opened the pull request:
# Pull Request Details

## What issue does this PR address

- Add APIs to stream data from / to POD and improve the parsing of RDF triples.

- Link to associated issue:
 - #249 
 - #254 

## Checklist

Complete the check-list below to ensure your branch is ready for PR.

Flutter Style Guide: https://survivor.togaware.com/gnulinux/flutter-style.html

- [ ] Screenshots included in linked issue
- [x] Changes adhere to the team style and coding guideline
- [x] No confidential information
- [x] No duplicated content
- [ ] No lint check errors related to your changes (`make prep` or `flutter analyze lib`)
- [ ] Pre-exisiting lint errors noted: [HERE]
- [x] Tested on at least one device
  - [ ] Android Phone
  - [ ] Android Emulator
  - [ ] Chrome on Android
  - [ ] Chrome
  - [ ] iOS
  - [x] Linux
  - [ ] MacOS
  - [x] Windows
- [x] Added 2 reviewers (or 1 for private repositories then they add another)

## Finalising

Once PR discussion is complete and 2 reviewers have approved:

- [x] Merge dev into the branch
- [x] Resolve any conflicts
- [x] Add one line summary into CHANGELOG.md
- [x] Bump appropriate version number in pubspec.yaml
- [x] Push to git repository and review
- [x] Merge PR into dev

cdawei committed:
Add send/post stream API

cdawei committed:
WIP

cdawei committed:
Refactor parseTTL for issue #254

cdawei committed:
Downgrade custom_lint to resolve version solving issue

cdawei committed:
Check duplicated values of triples for data encryption

cdawei committed:
WIP

cdawei committed:
Merge branch 'dc/254_refactor_parseTTL' into dc/249_stream_api

cdawei committed:
Add fetch() that returns a stream of file content

cdawei commented:
To test this pull request, please run the `demopod` app, click `Upload / Download Large File` button and upload, download, or delete a file as shown here https://github.com/anusii/solidpod/issues/186#issuecomment-2229890527

cdawei committed:
Improve error handling

cdawei committed:
Adapt code to use the updated readPod, use existing widget (for the child argument) instead of creating a new one.

srodriguez142857 commented:
For #249 (API to stream data), the app is working as expected.
I ran a test with the expected behaviour uploading, downloading, and deleting a large file (I used an MP4 video recording of over 109 MB long).

![image](https://github.com/user-attachments/assets/4563b03e-51fb-4988-b9a0-77cceeb8bf48)
![image](https://github.com/user-attachments/assets/fcf856f7-2b31-4683-a37a-8617bb7ce8f5)

All good for this function.

FYI: I got the following output in the terminal.

```
flutter: AuthDataManager => _getTokenResponse() refreshing expired token
flutter: AuthDataManager => _getTokenResponse() failed: OpenIdException(invalid_grant): grant request is invalid
flutter: AuthDataManager => getAccessToken() failed
flutter: solidAuthenticate() => checkLoggedIn() => false
flutter: solidAuthenticate() => solid_auth.authenticate(https://solid.dev.empwr.au)
flutter: server started 4400
flutter: AuthDataManager => saveAuthData() done
flutter: Permission granted: true
flutter: permissions-log.ttl
flutter: .acl
** flutter: Exception: Exception: Resource "exampleApp/data/key-value.ttl" does not exist.
** flutter: ACL file for "https://solid.dev.empwr.au/sergio101/exampleApp/data/large_file.bin.ttl" does not exist.
flutter: Deleted IndividualKeyRecord {
flutter:     filePath: exampleApp/data/large_file.bin.ttl,
flutter:     encIndKey: qF+IIIhgfMaVC6Hiwy1onspmH44m8seeTGkJAgtO09wb4WHy0qdPQzd0wpvYdV7t,
flutter:     iv: AqoI/IWGqY/iKJ3/YL870g==
flutter: }
flutter: Deleted large_file.bin
```
I'm unsure if the outputs indicated with `** ` are expected and correct.

---
For #254, I'm unsure on how to test it.
@cdawei, please let me know the details on how to test this from the app to check the expected TTL structure.
Thanks.

cdawei commented:
> For #249 (API to stream data), the app is working as expected. I ran a test with the expected behaviour uploading, downloading, and deleting a large file (I used an MP4 video recording of over 109 MB long).
> 
> ** flutter: Exception: Exception: Resource "exampleApp/data/key-value.ttl" does not exist.

Thanks Sergio. This exception is expected when the `key-value.ttl` does not exist, the app catches this exception and creates the `key-value.ttl` when new key-vaue pairs are submitted.

> ** flutter: ACL file for "https://solid.dev.empwr.au/sergio101/exampleApp/data/large_file.bin.ttl" does not exist.
> flutter: Deleted IndividualKeyRecord {
> flutter:     filePath: exampleApp/data/large_file.bin.ttl,
> flutter:     encIndKey: qF+IIIhgfMaVC6Hiwy1onspmH44m8seeTGkJAgtO09wb4WHy0qdPQzd0wpvYdV7t,
> flutter:     iv: AqoI/IWGqY/iKJ3/YL870g==
> flutter: }
> flutter: Deleted large_file.bin
> 
> I'm unsure if the outputs indicated with `** ` are expected and correct. 

This might be that the CSS server automatically deletes its ACL file when `large_file.bin.ttl` is deleted, which seems like a new feature of recent versions of CSS. I created a new issue #257 to test this. 

> For #254, I'm unsure on how to test it. @cdawei, please let me know the details on how to test this from the app to check the expected TTL structure. Thanks.

The upload/download/delete functions use the updated `parseTTL()` so it should have already been tested. Additional testing could be clicking `Show POD Data File`, `Show Encryption Key`, and `Update Security Key` buttons and see if those functionalities will work as expected.

srodriguez142857 commented:
Thanks @cdawei.
I did another round of testing.  All good.

srodriguez142857 reviewed (approved):
The app is working as expected.

anushkavidanage reviewed (approved):
All working as expected. Thanks Dawei.

cdawei committed:
Update change log and bump version to 0.6.6
